File: Scores/Strategies/EnumSortingStrategy.cs
using Scores.Enums;
using Scores.Interfaces;
using Scores.Models;
using System;

namespace Scores.Strategies
{
    public class EnumSortingStrategy<TEnum> : IScoreSortingStrategy
        where TEnum : struct, Enum
    {
        private readonly Func<string, string>? _preprocessScore;

        public EnumSortingStrategy(Func<string, string>? preprocessScore = null)
        {
            _preprocessScore = preprocessScore;
        }

        public List<StudentScore> SortScores(List<StudentScore> scores)
        {
            var orderedScores = scores.OrderBy(s =>
            {
                string scoreValue = _preprocessScore != null ? _preprocessScore(s.Score) : s.Score;
                
                if (Enum.TryParse<TEnum>(scoreValue, true, out var enumValue))
                    return (int)(object)enumValue;
                return 0;
            });

            return orderedScores.Reverse().ToList();
        }
    }
}

File: scores-php/src/Strategies/EnumSortingStrategy.php
<?php

namespace Scores\Strategies;

use Scores\Interfaces\IScoreSortingStrategy;
use Scores\Models\StudentScore;

abstract class EnumSortingStrategy implements IScoreSortingStrategy
{
    protected string $enumClass;
    
    public function sortScores(array $scores): array
    {
        usort($scores, function(StudentScore $a, StudentScore $b) {
            $valueA = $this->getEnumValue($a->score);
            $valueB = $this->getEnumValue($b->score);
            
            return $valueB <=> $valueA;
        });
        
        return $scores;
    }
    
    protected function getEnumValue(string $score): int
    {
        $enumClass = $this->enumClass;

        foreach ($enumClass::cases() as $case) {
            if ($case->name === $score) {
                return $case->value;
            }
        }

        foreach ($enumClass::cases() as $case) {
            if (strcasecmp($case->name, $score) === 0) {
                return $case->value;
            }
        }

        if (is_numeric($score)) {
            try {
                $enum = $enumClass::from((int)$score);
                return $enum->value;
            } catch (\ValueError $e) {
            }
        }

        return 0;
    }
}

